src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,960 @@
+    if( !omt_path.empty() && veh_omt == omt_path.back() ) {
+        omt_path.pop_back();
+    }
+    if( omt_path.empty() ) {
+        stop_autodriving( false );
+        return autodrive_result::finished;
+    }
+    if( !active_autodrive_controller ) {
+        active_autodrive_controller = std::make_shared<autodrive_controller>( *this, driver );
+    }
+    if( &active_autodrive_controller->get_driver() != &driver ) {
+        debugmsg( "Driver changed while auto-driving" );
+        stop_autodriving();
+        return autodrive_result::abort;
+    }
+    cata::optional<navigation_step> next_step = active_autodrive_controller->compute_next_step();
+    if( !next_step ) {
+        // message handles pathfinding failure either due to obstacles or inability to see
+        driver.add_msg_if_player( _( "Can't see a path forward." ) );
+        stop_autodriving( false );
+        return autodrive_result::abort;
+    }
+    if( next_step->pos != veh_pos ) {
+        debugmsg( "compute_next_step returned an invalid result" );
+        stop_autodriving();
+        return autodrive_result::abort;
+    }
+    if( next_step->target_velocity == 0 && velocity == 0 ) {
+        stop_autodriving();
+        return autodrive_result::finished;
+    }
+    cruise_on = true;
+    cruise_velocity = next_step->target_velocity;
+
+    // check for collisions before we steer, since steering may end our turn
+    // which would cause the vehicle to move and maybe crash
+    switch( active_autodrive_controller->check_collision_zone( next_step->steering_dir ) ) {
+        case collision_check_result::no_visibility:
+            driver.add_msg_if_player( m_warning, _( "You can't see anything in front of your vehicle!" ) );
+            stop_autodriving();
+            return autodrive_result::abort;
+        case collision_check_result::close_obstacle:
+            driver.add_msg_if_player( m_warning, _( "You're about to crash into something!" ) );
+            stop_autodriving( false );
+            return autodrive_result::abort;
+        case collision_check_result::partial_visibility:
+        case collision_check_result::obstacle:
+            active_autodrive_controller->reduce_speed();
+            if( cruise_velocity > VMIPH_PER_TPS ) {
+                cruise_velocity = VMIPH_PER_TPS;
+            }
+            break;
+        case collision_check_result::ok:
+            break;
+    }
+
+    int turn_delta = orientation_diff( next_step->steering_dir, normalize_orientation( turn_dir ) );
+    // pldrive() does not handle steering multiple times in one call correctly
+    // call it multiple times, matching how a player controls the vehicle
+    for( int i = 0; i < std::abs( turn_delta ); i++ ) {
+        if( driver.moves <= 0 ) {
+            // we couldn't steer as many times as we wanted to but there's
+            // nothing we can do about it now, hope we don't crash!
+            break;
```

Comment:
Possibly reduce speed?

Reply:
It's too late for that. The player's turn is over; doing anything else now would be cheating.
The next call to do_autodrive() will have to deal with the consequences.

Reply:
I see that this is non-obvious, so added a comment.

Reply:
If there are 2+ steerings left, and driver.moves is not sufficient to do more than 1, then brake? This is also a way of making autodrive max speed better match avatar (or NPC driver) speed.

Reply:
The problem is that I don't know how much it'll cost to turn before calling plturn(). But if I could find out that the desired number of turns is not possible I would decrease max_steering and max_speed and try to find another path, rather than brake.

Reply:
Right now it doesn't matter because max_steering is capped at 1. The moves check is just "defensive programming".

Reply:
I should also mentioned that I do plan to eventually make some changes to the vehicle code to be able to accurately predict the move cost of steering, as part of the work to enable max_steering > 1. Just not in this PR.